File: rocketmq-client4cpp/src/ClientConfig.cpp
/**
 * Copyright (C) 2010-2013 kangliqiang, kangliq@163.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <stdlib.h>
#include <sstream>

#include "MQClientException.h"
#include "SocketUtil.h"
#include "ClientConfig.h"
#include "UtilAll.h"
#include "MixAll.h"

namespace rmq
{

ClientConfig::ClientConfig()
{
    char* addr = getenv(MixAll::NAMESRV_ADDR_ENV.c_str());
    if (addr)
    {
        m_namesrvAddr = addr;
    }
    else
    {
        m_namesrvAddr = "";
    }

    m_clientIP = getLocalAddress();
    m_instanceName = "DEFAULT";
    m_clientCallbackExecutorThreads = UtilAll::availableProcessors();
    m_pollNameServerInterval = 1000 * 30;
    m_heartbeatBrokerInterval = 1000 * 30;
    m_persistConsumerOffsetInterval = 1000 * 5;
}

ClientConfig::~ClientConfig()
{
}

std::string ClientConfig::buildMQClientId()
{
    return m_clientIP + "@" + m_instanceName;
}

void ClientConfig::changeInstanceNameToPID()
{
    if (m_instanceName == "DEFAULT")
    {
        m_instanceName = UtilAll::toString(UtilAll::getPid());
    }
}


void ClientConfig::resetClientConfig(const ClientConfig& cc)
{
    m_namesrvAddr = cc.m_namesrvAddr;
    m_clientIP = cc.m_clientIP;
    m_instanceName = cc.m_instanceName;
    m_clientCallbackExecutorThreads = cc.m_clientCallbackExecutorThreads;
    m_pollNameServerInterval = cc.m_pollNameServerInterval;
    m_heartbeatBrokerInterval = cc.m_heartbeatBrokerInterval;
    m_persistConsumerOffsetInterval = cc.m_persistConsumerOffsetInterval;
}

ClientConfig ClientConfig::cloneClientConfig()
{
    return *this;
}

std::string ClientConfig::getNamesrvAddr()
{
    return m_namesrvAddr;
}

void ClientConfig::setNamesrvAddr(const std::string& namesrvAddr)
{
    m_namesrvAddr = namesrvAddr;
}

std::string ClientConfig::getClientIP()
{
    return m_clientIP;
}

void ClientConfig::setClientIP(const std::string& clientIP)
{
    m_clientIP = clientIP;
}

std::string ClientConfig::getInstanceName()
{
    return m_instanceName;
}

void ClientConfig::setInstanceName(const std::string& instanceName)
{
    m_instanceName = instanceName;
}

int ClientConfig::getClientCallbackExecutorThreads()
{
    return m_clientCallbackExecutorThreads;
}

void ClientConfig::setClientCallbackExecutorThreads(int clientCallbackExecutorThreads)
{
    m_clientCallbackExecutorThreads = clientCallbackExecutorThreads;
}

int ClientConfig::getPollNameServerInterval()
{
    return m_pollNameServerInterval;
}

void ClientConfig::setPollNameServerInterval(int pollNameServerInterval)
{
    m_pollNameServerInterval = pollNameServerInterval;
}

int ClientConfig::getHeartbeatBrokerInterval()
{
    return m_heartbeatBrokerInterval;
}

void ClientConfig::setHeartbeatBrokerInterval(int heartbeatBrokerInterval)
{
    m_heartbeatBrokerInterval = heartbeatBrokerInterval;
}

int ClientConfig:: getPersistConsumerOffsetInterval()
{
    return m_persistConsumerOffsetInterval;
}

void ClientConfig::setPersistConsumerOffsetInterval(int persistConsumerOffsetInterval)
{
    m_persistConsumerOffsetInterval = persistConsumerOffsetInterval;
}


std::string ClientConfig::toString() const
{
	std::stringstream ss;
	ss 	<< "{namesrvAddr=" << m_namesrvAddr
		<< ",clientIP=" << m_clientIP
		<< ",instanceName=" << m_instanceName
		<< ",clientCallbackExecutorThreads=" << m_clientCallbackExecutorThreads
		<< ",pollNameServerInteval=" << m_pollNameServerInterval
		<< ",heartbeatBrokerInterval=" << m_heartbeatBrokerInterval
		<< ",persistConsumerOffsetInterval=" << m_persistConsumerOffsetInterval
		<<"}";
	return ss.str();
}


}

File: rocketmq-go/model/config/client_config.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package config

import (
	"bytes"
	"time"
)

// client common config
type ClientConfig struct {
	nameServerAddress string // only this is in use

	clientIP                      string
	instanceName                  string
	clientCallbackExecutorThreads int // TODO: clientCallbackExecutorThreads
	// Pulling topic information interval from the named server
	pullNameServerInterval time.Duration // default 30
	// Heartbeat interval in microseconds with message broker
	heartbeatBrokerInterval time.Duration // default 30
	// Offset persistent interval for consumer
	persistConsumerOffsetInterval time.Duration // default 5
	unitMode                      bool
	unitName                      string
	vipChannelEnabled             bool
}

func NewClientConfig() *ClientConfig {
	return &ClientConfig{
		unitMode:                      false,
		pullNameServerInterval:        time.Second * 30,
		heartbeatBrokerInterval:       time.Second * 30,
		persistConsumerOffsetInterval: time.Second * 30,
	}
}

func (config *ClientConfig) BuildMQClientId() string {
	var buffer bytes.Buffer
	buffer.WriteString(config.clientIP)
	buffer.WriteString("@")
	buffer.WriteString(config.instanceName)
	if config.unitName != "" {
		buffer.WriteString("@")
		buffer.WriteString(config.unitName)
	}
	return buffer.String()
}

func (config *ClientConfig) ChangeInstanceNameToPID() {
	// TODO
}

func (config *ClientConfig) ResetClientConfig(cfg *ClientConfig) {
	// TODO
}

func (config *ClientConfig) CloneClientConfig() *ClientConfig {
	return &ClientConfig{
		nameServerAddress:             config.nameServerAddress,
		clientIP:                      config.clientIP,
		instanceName:                  config.instanceName,
		clientCallbackExecutorThreads: config.clientCallbackExecutorThreads,
		pullNameServerInterval:        config.pullNameServerInterval,
		heartbeatBrokerInterval:       config.heartbeatBrokerInterval,
		persistConsumerOffsetInterval: config.persistConsumerOffsetInterval,
		unitMode:                      config.unitMode,
		unitName:                      config.unitName,
		vipChannelEnabled:             config.vipChannelEnabled,
	}
}

func (config *ClientConfig) ClientIP() string {
	return config.clientIP
}

func (config *ClientConfig) SetClientIP(s string) {
	config.clientIP = s
}

func (config *ClientConfig) InstanceName() string {
	return config.instanceName
}

func (config *ClientConfig) SetInstanceName(s string) {
	config.instanceName = s
}

func (config *ClientConfig) NameServerAddress() string {
	return config.nameServerAddress
}

func (config *ClientConfig) SetNameServerAddress(s string) {
	config.nameServerAddress = s
}

func (config *ClientConfig) ClientCallbackExecutorThreads() int {
	return config.clientCallbackExecutorThreads
}

func (config *ClientConfig) SetClientCallbackExecutorThreads(threads int) {
	config.clientCallbackExecutorThreads = threads
}

func (config *ClientConfig) PullNameServerInteval() time.Duration {
	return config.pullNameServerInterval
}

func (config *ClientConfig) SetPullNameServerInteval(interval time.Duration) {
	config.pullNameServerInterval = interval
}

func (config *ClientConfig) HeartbeatBrokerInterval() time.Duration {
	return config.heartbeatBrokerInterval
}

func (config *ClientConfig) SetHeartbeatBrokerInterval(interval time.Duration) {
	config.heartbeatBrokerInterval = interval
}

func (config *ClientConfig) PersistConsumerOffsetInterval() time.Duration {
	return config.persistConsumerOffsetInterval
}

func (config *ClientConfig) SetPersistConsumerOffsetInterval(interval time.Duration) {
	config.persistConsumerOffsetInterval = interval
}

func (config *ClientConfig) UnitName() string {
	return config.unitName
}

func (config *ClientConfig) SetUnitName(name string) {
	config.unitName = name
}

func (config *ClientConfig) UnitMode() bool {
	return config.unitMode
}

func (config *ClientConfig) SetUnitMode(mode bool) {
	config.unitMode = mode
}

func (config *ClientConfig) VipChannelEnabled() bool {
	return config.vipChannelEnabled
}

func (config *ClientConfig) SetVipChannelEnabled(enable bool) {
	config.vipChannelEnabled = enable
}

func (config *ClientConfig) String() string {
	//TODO
	return ""
}
